mlflow/models/model.py

```diff
@@ -223,6 +223,10 @@ def add_flavor(self, name, **params):
         self.flavors[name] = params
         return self
 
+    def add_metadata(self, metadata):
+        """Add metadata to the model"""
+        self.__dict__.update(metadata)
```

Comment:
I'm not the biggest fan of this and don't like obfuscated mutation of object members. 
This should be a property so that it can have a descriptor access and should have an accompanying setter to provide assignment. The overriding of `__dict__` key entries can wreak all sorts of havok for users who specify an in-use key in the dictionary that would result in an invalidated and corrupted object that gets saved or logged. 

See the implementation for `signature` for guidance.

Reply:
We wanted to avoid modifying too much code in the current implementation, but you're right, this isn't ideal.

I've updated the code with an actual property.

What I don't like is that now the *metadata* keyword is repeated when accessing the values.
```python
model = mlflow.pyfunc.load_model(model_uri="models:/MyModel/latest")
print(model.metadata.metadata["metadata_key"])
print(model.metadata.get_model_info().metadata["metadata_key"])
# or
model_info = mlflow.models.get_model_info(model_uri="models:/MyModel/latest")
print(model_info.metadata["stateful"])
```

It is not a major problem though.  
It could be avoided by using adding a `__getattr__` method to both `Model` and `ModelInfo` classes:
```python
def __getattr__(self, attr):
    if attr not in self.__dict__:
        return self.metadata[attr]
    return super().__getattr__(attr)
```

Resulting in direct access to the metadata values:
```python
model = mlflow.pyfunc.load_model(model_uri="models:/MyModel/latest")
print(model.metadata.metadata_key)
print(model.metadata.get_model_info().metadata_key)
# or
model_info = mlflow.models.get_model_info(model_uri="models:/MyModel/latest")
print(model_info.metadata_key)
```

Another solution is to change the naming, but I couldn't come up with anything better.